onnxruntime/python/tools/automl/data_frame_tool.py

```diff
@@ -0,0 +1,137 @@
+        :param sess_options: see onnxruntime.SessionsOptions
+        """
+        self._model_path = model_path
+        self._sess_options = sess_options
+
+    def _process_input_list(self, df, input_metas, require):
+        "Return a dictionary of input_name : a typed and shaped np.array of values for a given input_meta"
+        """
+        The function does the heavy lifting for _get_input_feeds()
+
+        :param df: See :class:`pandas.DataFrame`. 
+        :param input_metas: a list of name/type pairs
+        :require is a boolean. If True this helper throws on a missing input.
+        
+        """
+        feeds = {}
+        # Process mandadory inputs. Raise an error if anything is not present
+        for input_meta in input_metas:
+            shape = [dim if dim else 1 for dim in input_meta.shape]
+            # We fully expect all the types are in the above dictionary
+            assert input_meta.type in types_dict
+            print("Processing input: " + input_meta.name)
+            if input_meta.name in df.columns:
+                expected_type = types_dict[input_meta.type]
+                # float16 and bool will always require exact match
+                # We attempt to convert any type to a string if it is required.
+                # With strings we always want to put this into a flat array, cast to np.object and then reshape as object
+                if input_meta.type == 'tensor(string)':
+                    #print('Col: {} processed as string type: {} '.format(input_meta.name, df[input_meta.name].dtype))
+                    feeds[input_meta.name] = np.array([df[input_meta.name][0]]).astype(expected_type).reshape(shape)
+                elif expected_type == df[input_meta.name].dtype: # If there is an exact match we take as is
+                    #print('Col: {} processed exact match type: {} '.format(input_meta.name, df[input_meta.name].dtype))
+                    feeds[input_meta.name] = np.array([df[input_meta.name][0]]).astype(expected_type).reshape(shape)
+                elif expected_type in ort_float_set and str(df[input_meta.name].dtype) in pd_float_set:
+                    #print('Col: {} processed as floating type: {} '.format(input_meta.name, df[input_meta.name].dtype))
+                    feeds[input_meta.name] = np.array([df[input_meta.name][0]]).astype(expected_type).reshape(shape)
+                elif expected_type in ort_int_set and str(df[input_meta.name].dtype) in pd_int_set:
+                    #print('Col: {} processed as integer type: {} '.format(input_meta.name, df[input_meta.name].dtype))
+                    feeds[input_meta.name] = np.array([df[input_meta.name][0]]).astype(expected_type).reshape(shape)
+                else:
+                    raise TypeError("Input {} expected to be of type: {} got {} ".format(
+                                input_meta.name, expected_type, df[input_meta.name].dtype))
+            elif require:
```

Comment:
I'm having trouble wrapping my head around the scenario where require is set to False. In that case, ORT would generate an exception because the input wasn't provided wouldn't it? If so, maybe we should always raise an exception.

Reply:
I see that the code below is invoking the method with require set to False when processing the initializers. Maybe add a comment near that invocation that mentions why errors are not fatal in that scenario.

Reply:
We should always raise exception when required input is not provided. There is not a way around this. We will chat offline. You can also talk to Narish as he requested overridable initializers to be exposed.

---
In reply to: [351396003](https://github.com/microsoft/onnxruntime/pull/2456#discussion_r351396003) [](ancestors = 351396003)